src/cressi_goa.c

```diff
@@ -377,6 +382,117 @@ cressi_goa_device_set_fingerprint (dc_device_t *abstract, const unsigned char da
 	return DC_STATUS_SUCCESS;
 }
 
+struct cressi_goa_irda_api_conf {
+	unsigned char version;
+	unsigned int idlen;
+	unsigned char logbook_entry_len;
+	unsigned int logbook_fp_offset;
+	unsigned char cmd_logbook;
+};
+
+static int
+cressi_goa_irda_api_version(dc_device_t *abstract, dc_event_devinfo_t *devinfo)
+{
+	if (devinfo->model > 11) {
+		ERROR (abstract->context, "Unknown model %d.", devinfo->model);
+		return -1;
+	}
+	int version;
+	if (devinfo->firmware >= 161 && devinfo->firmware <= 165) {
+		version = 0;
+	} else if (devinfo->firmware >= 166 && devinfo->firmware <= 169) {
+		version = 1;
+	} else if (devinfo->firmware >= 170 && devinfo->firmware <= 179) {
+		version = 2;
+	} else if ((devinfo->firmware >= 100 && devinfo->firmware <= 110)
+			|| devinfo->firmware == 900) {
+		version = 3;
+	} else if (devinfo->firmware >= 200 && devinfo->firmware <= 205) {
+		version = 4;
+	} else {
+		ERROR (abstract->context, "Unknown firmware version %d.", devinfo->firmware);
+		return -1;
+	}
+	const bool version_support_on_model[5][11] = {
+		/*  1      2      3      4      5      6      7      8      9      10     11  */
+		{  true,  true, false, false, false, false, false, false, false, false, false }, /* API v0 */
+		{  true,  true, false, false, false, false, false, false, false, false, false }, /* API v1 */
+		{  true,  true, false, false, false, false, false, false,  true, false, false }, /* API v2 */
+		{  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true }, /* API v3 */
+		{  true,  true, false,  true,  true, false, false, false,  true,  true, false }, /* API v4 */
+	};
+	if (!version_support_on_model[version][devinfo->model - 1]) {
+		ERROR (abstract->context, "Firmware version %d of Model %d not known to have support for API v%d.",
+				devinfo->firmware, devinfo->model, version);
+		return -1;
+	}
+	return version;
+}
+
+static dc_status_t
+cressi_goa_read_id(dc_device_t *abstract, bool emit_events, struct cressi_goa_irda_api_conf *conf)
+{
+	static const struct cressi_goa_irda_api_conf version_conf[] = {
+		{ 0,  9, SZ_HEADER, FP_OFFSET, CMD_LOGBOOK },
+		/*    4 is the new version
+		 *   11 is the new response length to the `CMD_VERSION` command
+		 *   15 is the length of an entry in the Logbook
+		 *    3 is the offset to the START date in the Logbook header
+		 * 0x23 is the new command to request the Logbook
+		 */
+		{ 4, 11,        15,         3, CMD_LOGBOOK_V4 },
```

Comment:
Am I right that the difference in the size of the version response only allows to distinguish between V4 and the other versions, but not which of those other versions (V0 to V3)? So basically we can use this version size to know whether we need to send the CMD_LOGBOOK or CMD_LOGBOOK_V4 (and the size and format of the logbook entries). During the download, that's also the only info we really need here.

Or is it possible to have the 11 byte version (e.g. V4) here, and get back some different version from cressi_goa_irda_api_version?

The actual API version, which is obtained in cressi_goa_irda_api_version (from the firmware version and the table), is only needed later during parsing? If that's correct, then we shouldn't validate this info here. If we can download the dive but not parse them, that should not cause the download to fail, only the parsing. Ideally the download code shouldn't have to know the data format of the dives (or at least as little as possible). So I think this logic should be moved to the parser. Since the firmware version isn't available in the dive header, that means we mus pass that info somehow. The most obvious option is probably to prepend the entire version data to each dive instead of just the version number we extracted from this version data? That way we can push all the logic that needs to deal with the different data formats to the parser, where it belongs.

Reply:
> Am I right that the difference in the size of the version response only allows to distinguish between V4 and the other versions, but not which of those other versions (V0 to V3)?

Correct.

> So basically we can use this version size to know whether we need to send the CMD_LOGBOOK or CMD_LOGBOOK_V4 (and the size and format of the logbook entries).

Yes.

> Or is it possible to have the 11 byte version (e.g. V4) here, and get back some different version from cressi_goa_irda_api_version?

TBH I don't know. The only thing I can say for sure is that the firmware version that is returned via the ID command is mapped to the API version. Whether there's a firmware that replies with a v4 ID, but a firmware version indicating that its real version is <v4 I can't say, but I doubt it :)

> The most obvious option is probably to prepend the entire version data to each dive instead of just the version number we extracted from this version data? That way we can push all the logic that needs to deal with the different data formats to the parser, where it belongs.

Sure, we could move that logic to the parser.

I'd suggest to adapt the output format of the downloaded data to:

`0xdc 0xdc <1 octet version=0> <1 octet idlen> <9 or 11 octet id> <23 or 15 octet logbook entry> <dive data>`

The `version` would be a "libdivecomputer version field" if we ever have to change that format.

Reply:
Done.